### ISSUE TITLE
HTTP Client benchmarks

### ISSUE THREAD

**Issue by grigi** [vibora-io/benchmarks#2]
Since Vibora has a HTTP client as well, we need representative benchmarks to compare it to other Python clients.

I have started on a set of benchmarks for the sync python HTTP clients.
I tested against Apache2 because I can disable pipelining, so that we can confirm that pipelining is working as expected.
I used WRK as a reference benchmark.
I limited all benchmarks to not use more than a single core (to keep the benchmark more real)(and I only have 2 cores, so can only effectively dedicate a single core to the app).
I also compared with/without using sessions(or object reuse).

So far my results are:

Client| HTTP/1.0 | HTTP/1.0 + session | HTTP/1.1 | HTTP/1.1 + session
------|-----------|---------------------|----------|---------------------
Requests|555.91|605.80|531.86|755.34
urllib3|1295.25|1410.67|1220.07|1582.52
PyCurl|4136.61|4810.15|3992.62|7965.60
WRK 1 connection||5840.48||15082.47
WRK 10 connections||12045.53||32965.53

This is on my i5-6300U notebook, running linux. Benchmark is Py3.6.5 on Alpine Linux in a local Docker.

I'll add async tests, but it is past midnight now.

I'll also do a PR,  but how do you want to have this all integrated, as my style of doing docker is vastly different than yours? Should we have sub-projects in here, each with their own docker to run tests?
e.g. `HTTP Server`/`HTTP Client`/`Template`/`Schema` ?

**Comment by grigi** [comment]
Also, this was just benchmarked against HTTP, still plan to add HTTPS + certificate validation (as that is a very real use-case)

**Comment by grigi** [comment]
Woah, Python3.7b5 has some really nice speedups:

Client| HTTP/1.0 | HTTP/1.0 + session | HTTP/1.1 | HTTP/1.1 + session
------|-----------|---------------------|----------|---------------------
Requests|600.63|695.52|587.73|791.97
urllib3|1337.56|1473.05|1254.08|2011.80
PyCurl|4145.22|4741.74|4247.15|10178.16

**Comment by grigi** [comment]
Added AioHTTP: (on Py3.6)

Client| No Pipelining | No Pipelining +  session | No Pipelining + session + uvloop | Pipelined | Pipelined + session | Pipelined + session + uvloop
------|--------------|---------------------------|-----------------------------------|----------|------------------------|------------------------------
AioHTTP 1 connection|611.17|732.51|967.04|651.23|1413.81|1566.37
AioHTTP 10 connections|974.12|996.07|1333.39|984.63|1733.25|1874.67

AioHTTPs performance is about in line with urllib3, but way behind PyCurl.

**Comment by grigi** [comment]
Added `asks` + `trio`:

Client| No Pipelining | No Pipelining +  session | Pipelined | Pipelined + session 
------|--------------|---------------------------|----------|---------------------
asks+trio 1 connection|518.92|503.89|529.89|1258.48
asks+trio 10 connections|665.35|671.64|655.35|1451.86

A bit slower than AioHTTP across the board.

**Comment by grigi** [comment]
Added `vibora.client` I had to look at the source. Also the way sessions work is a bit suspect. When I define a session, exactly as in the code, it runs slower...
And the session object creation takes ~10ms on my system just by itself...

Client| No Pipelining | Pipelined
------|--------------|----------
vibora 1 connection|1776.74|2839.16
vibora 10 connections|3826.65|4084.55

Congratulations @frnkvieira it is easily the second-fastest Python implementation already, and the fastest async client I could find!
